tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SNS/ContextPropagation.cs

```diff
@@ -0,0 +1,84 @@
+// <copyright file="ContextPropagation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using Datadog.Trace.DataStreamsMonitoring;
+using Datadog.Trace.Propagators;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.SNS
+{
+    internal static class ContextPropagation
+    {
+        private const string SnsKey = "_datadog";
+
+        private static void Inject<TMessageRequest>(SpanContext context, IDictionary messageAttributes)
+        {
+            // Consolidate headers into one JSON object with <header_name>:<value>
+            var sb = Util.StringBuilderCache.Acquire(Util.StringBuilderCache.MaxBuilderSize);
+            sb.Append('{');
+            SpanContextPropagator.Instance.Inject(context, sb, default(StringBuilderCarrierSetter));
+            sb.Remove(startIndex: sb.Length - 1, length: 1); // Remove trailing comma
+            sb.Append('}');
+
+            var resultString = Util.StringBuilderCache.GetStringAndRelease(sb);
+            messageAttributes[SnsKey] = CachedMessageHeadersHelper<TMessageRequest>.CreateMessageAttributeValue(resultString);
+        }
+
+        public static void InjectHeadersIntoMessage<TMessageRequest>(IContainsMessageAttributes carrier, SpanContext spanContext)
```

Comment:
nit: where possible, we shouldn't pass ducktyped interfaces as method parameters directly, because it can cause boxing of structs. A better approach is to use generic type with an interface constraint, e.g.

This is 
```suggestion
        public static void InjectHeadersIntoMessage<TMessageRequest, TCarrier>(TCarrier carrier, SpanContext spanContext)
            where TCarrier : IContainsMessageAttributes
```

Note that this is currently an issue in the SQS version you based it on, so can push it back to a separate PR I think 🙂